src/ILCompiler.Build.Tasks/src/ComputeManagedAssemblies.cs

```diff
@@ -0,0 +1,142 @@
+        /// </summary>
+        [Required]
+        public ITaskItem[] FrameworkAssemblies
+        {
+            get;
+            set;
+        }
+
+        [Output]
+        public ITaskItem[] ManagedAssemblies
+        {
+            get;
+            set;
+        }
+
+        [Output]
+        public ITaskItem[] AssembliesToSkipPublish
+        {
+            get;
+            set;
+        }
+
+        public override bool Execute()
+        {
+            var list = new List<ITaskItem>();
+            var assembliesToSkipPublish = new List<ITaskItem>();
+
+            ITaskItem[] assemblies = this.Assemblies;
+
+            var coreRTFrameworkAssembliesToUse = new HashSet<string>();
+
+            foreach (var x in SdkAssemblies)
+            {
+                coreRTFrameworkAssembliesToUse.Add(Path.GetFileName(x.ItemSpec));
+            }
+
+            foreach (var x in FrameworkAssemblies)
+            {
+                coreRTFrameworkAssembliesToUse.Add(Path.GetFileName(x.ItemSpec));
+            }
+
+            for (int i = 0; i < assemblies.Length; i++)
+            {
+                ITaskItem taskItem = assemblies[i];
+
+                // Skip crossgen images
+                if (taskItem.ItemSpec.EndsWith(".ni.dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                // Skip the native apphost (whose name ends up colliding with the CoreRT output binary) and supporting assemblies
+                if (taskItem.ItemSpec.EndsWith("apphost.exe", StringComparison.OrdinalIgnoreCase) || taskItem.ItemSpec.Contains("hostfxr") || taskItem.ItemSpec.Contains("hostpolicy"))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                // Prototype aid - remove the native CoreCLR runtime pieces from the publish folder
+                if (taskItem.ItemSpec.Contains("microsoft.netcore.app") && (taskItem.ItemSpec.Contains("\\native\\") || taskItem.ItemSpec.Contains("/native/")))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                // Remove any assemblies whose implementation we want to come from CoreRT's package.
+                // Currently that's System.Private.* SDK assemblies and a bunch of framework assemblies.
+                if (coreRTFrameworkAssembliesToUse.Contains(Path.GetFileName(taskItem.ItemSpec)))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                try
+                {
+                    using (var moduleStream = File.OpenRead(taskItem.ItemSpec))
+                    using (var module = new PEReader(moduleStream))
+                    {
+                        var moduleMetadataReader = module.GetMetadataReader();
```

Comment:
Do we need to check `module.HasMetadata` first? (`GetMetadataReader` will throw if there's no metadata.)